docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1038 @@
+type ControlPlaneProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ControlPlaneProviderSpec   `json:"spec,omitempty"`
+   Status ControlPlaneProviderStatus `json:"status,omitempty"`
+}
+
+// ControlPlaneProviderSpec defines the desired state of ControlPlaneProvider
+type ControlPlaneProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// ControlPlaneProviderStatus defines the observed state of ControlPlaneProvider
+type ControlPlaneProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+
+// InfrastructureProvider is the Schema for the InfrastructureProviders API
+type InfrastructureProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   InfrastructureProviderSpec   `json:"spec,omitempty"`
+   Status InfrastructureProviderStatus `json:"status,omitempty"`
+}
+
+// InfrastructureProviderSpec defines the desired state of InfrastructureProvider
+type InfrastructureProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// InfrastructureProviderStatus defines the observed state of InfrastructureProvider
+type InfrastructureProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+```
+
+Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
+shared among all the provider types - Core, Bootstrap, ControlPlane, and
+Infrastructure.
+
+```golang
+// ProviderSpec defines the desired state of the Provider.
+type ProviderSpec struct {
+   // Version indicates the provider version.
+   // +optional
+   Version string `json:"version,omitempty"`
+
+   // Manager defines the properties that can be enabled on the controller manager for the provider.
+   // +optional
+   Manager ManagerSpec `json:"manager,omitempty"`
+
+   // Deployment defines the properties that can be enabled on the deployment for the provider.
+   // +optional
+   Deployment DeploymentSpec `json:"deployment,omitempty"`
+
+   // SecretName is the name of the Secret providing the configuration
+   // variables for the current provider instance, like e.g. credentials.
+   // Such configurations will be used when creating or upgrading provider components.
+   // +optional
+   SecretName *string
+
+   // FetchConfig determines how the operator will fetch the components and metadata for the provider.
+   // If nil, the operator will try to fetch components according to default
+   // settings embedded in the operator and in clusterctl.
+   // +optional
+   FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`
+
+   // Paused prevents the operator from reconciling the provider. This can be
+   // used when doing an upgrade or move action manually.
+   // +optional
+   Paused bool `json:"paused,omitempty"`
+}
+
+// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.
+type ManagerSpec struct {
```

Comment:
I've done this for ProfilerAddress and FeatureGates. 

It seems that there currently aren't any flags for Verbosity so the provider controller managers will need to include support for a verbosity flag. (_unless I'm missing something_)

Also I realized that for `MaxConcurrentReconciles`, we will need to have logic somewhere (either in the operator or the individual controllers) that can set this value to the various concurrency flags.
That is, for example, for the CAPI provider, this value will need to be set on `cluster-concurrency`, `machine-concurrency`, `machineset-concurrency`, `machinedeployment-concurrency`, `machinepool-concurrency` etc..
If we do this in the operator, that would mean the operator would need to have knowledge of these flags per provider (_which seems like a bad coupling of concerns_). 

We might have to support this flag for each provider and do this within each controller. WDYT?

Reply:
Another option would be to remove `MaxConcurrentReconciles` from the API and instead just have users provide the flags as container args if they want to specify a particular `*controller-concurrency` flag.

This would avoid having the need for extra logic in extrapolating how this property maps to a specific provider.
That is, for the `CoreProvider`,  `MaxConcurrentReconciles` needs to map to the above flags but for an `InfrastructureProvider` such as CAPA, it needs to map to `awscluster-concurrency` and `awsmachine-concurrency`.

Personally, I'm leaning to removing this field for the initial implementation and we can add it in if we see the need for an abstraction like this. 
WDYT?

Reply:
👍🏽  If there is a clear user need for something like this, then we should include it. 

Would it be acceptable for the provider controller managers to support a generic `--controller-concurrency` flag that if set, all the other flags such as `awscluster-concurrency` and `awsmachine-concurrency` would be set to that value. This would avoid having the operator to be coupled to the providers. That is, if users set `MaxConcurrentReconciles: 20`, the operator would add `--controller-concurrency=20` to the provider flags.